### PR TITLE
Recommendation algo

### PR DESCRIPTION
**Pull request by Fnuworsu**
# Description  
- What does this PR do?  
  `Implements a complete social follow/unfollow relationship system for the your-pace.ai community platform, enabling users to build social connections and curate their content experience.`  
- Why is it necessary or valuable?  
  `This feature transforms the platform from a simple content sharing tool into a social learning community. Users can now follow other learners, discover content from people they're interested in, and build meaningful connections around shared learning topics. This addresses the core need for social interaction and personalized content discovery in educational platforms.`  
---

# Resources and References  
- Tutorials, documentation, or code examples used:  
  `Prisma many-to-many relationship documentation for self-referencing User model`  
---

# Test Plan  
- How was this tested?  
  `1. Manual testing of all follow/unfollow scenarios through the UI
    2. Tested edge cases: self-follow prevention, duplicate follow prevention
    3. Verified database relationships and cascading behavior
    4. Tested all feed types (Followers, Following, All Users) with different user states
    5. Verified proper button states (Follow, Unfollow, Follow Back) in different contexts`  
- Screenshots or recordings:  
  `
Screenshot 2025-07-11 at 6 03 24 PM
Screenshot 2025-07-11 at 6 03 30 PM
Screenshot 2025-07-11 at 6 03 36 PM
Screenshot 2025-07-11 at 6 03 42 PM
`  
---

### CODE REVIEW

**Comment by Fnuworsu** [comment]
@dorcheng @sinap-fb update PR
# Description  
- What does this PR do?  
  `Implements a clean, efficient recommendation algorithm for the "For You" feed that replaces magic numbers with well-named constants. The algorithm scores posts based on category relevance, engagement
  metrics, user following relationships, trending analysis, content discovery, and recency factors.`  
- Why is it necessary or valuable?  
  `Provides users with personalized content recommendations instead of chronological feeds. The algorithm is maintainable and configurable through named constants, making it easy to tune recommendation
  weights. Significantly improves user engagement by surfacing relevant content based on multiple scoring factors.`  
- What is intentionally left out and will be done in a later PR?  
  `User behavior tracking and learning`

---

# Milestones / Related Work  
- Feature or user story this contributes to: `Technical Challenge #1`  
- Milestone or version target: `Smart Content Recommendations`  

---

# Resources and References  
- Tutorials, documentation, or code examples used:  
  `Social media recommendation patterns (Twitter, LinkedIn)`  
---

# Test Plan  
- How was this tested?  
  `- Manual testing with various post combinations and user following patterns
    - Console logging to verify scoring calculations
    - Database integration testing with real post data
    - Performance testing with different dataset sizes
    - Algorithm weight tuning through manual testing`  
## Edge Cases Tested  
- ` - Posts with zero engagement (likes = 0, comments = 0)
  - Very new posts (hoursOld < 0.1)
  - Posts from new creators (followerCount < 10)
  - Users with no following relationships
  - Missing category information (defaults to 'OTHER')`

---

# Additional Notes  
- Known limitations: ` - Algorithm weights are manually tuned (no ML optimization)
  - No user behavior learning or adaptation
  - Performance scales linearly with dataset size
  - Requires post engagement data to be effective`  
- Planned follow-up PRs: `User interaction tracking for algorithm improvement`  
- Reviewer requests or feedback needed: `Algorithm weight balance evaluation`

File: frontend/src/components/MediaSection/Flashcards.jsx [hunk 1]
```diff
@@ -1,96 +1,110 @@
...
-                        <p>{flashcards[currentIndex].answer}</p>
-                        <span className="flip-hint">Click to flip back</span>
+            {loading ? (
+                <div className="flashcards-container">Loading flashcards...</div>
+            ) : flashcards.length === 0 ? (
+                <div className="flashcards-container">No flashcards available.</div>
+            ) : (
+                <>
+                    <div className="flashcard-wrapper">
+                        <div
```
Comment (sinap-fb) [inline frontend/src/components/MediaSection/Flashcards.jsx:72]: as a stretch goal, let's try to make this accessible.  One way is to use a `button` element, but then you'll need to adjust the styling probably.

File: agents/server.js [hunk 2]
```diff
@@ -97,12 +179,94 @@ app.post("/api/quiz", async (req, res) => {
...
+           `${chapter.title}: ${chapter.summary}`
+       ).join('\n\n')
+
+       const response = await ai.models.generateContent({
+           model: 'gemini-2.0-flash-001',
+           contents: `These are the chapter summaries: ${chaptersText}. Do this: ${CHAPTERS_SUMMARY_PROMPT}`
+       })
+
+       res.json({ summary: response.text.trim() })
+   } catch (error) {
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply agents/server.js:224]: @sinap-fb that's true. I will do ``throw new Error({ cause: error })``

File: agents/server.js [hunk 3]
```diff
@@ -97,12 +179,94 @@ app.post("/api/quiz", async (req, res) => {
...
+
+
+  app.post("/api/quiz-from-chapters", async (req, res) => {
+    try {
+        const { body: { chapters } } = req
+        if (!chapters || chapters.length === 0) return res.status(400).json({ error: 'No chapters provided' })
+
+        const chaptersText = chapters.map(chapter =>
+            `${chapter.title}: ${chapter.summary}`
+        ).join('\n\n')
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply agents/server.js:257]: @sinap-fb yeah, that makes a lot of sense 👍

File: backend/recommendationAlgo/recommendation.js [hunk 4]
```diff
@@ -1,34 +1,113 @@
...
+   'LANGUAGE': 0.8,
+   'ARTS_CRAFTS': 0.7,
+   'COOKING': 0.7,
+   'MUSIC': 0.6,
+   'ENTERTAINMENT': 0.6,
+   'OTHER': 0.4
 }
 
-const priorityQueue = new PriorityQueue((a,b) => a.score > b.score)
+const MIN_HOURS_TO_PREVENT_DIVISION = 0.1
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/recommendationAlgo/recommendation.js:24]: @yes. It's a bit scrappy but that was what I could do at that time

File: backend/recommendationAlgo/recommendation.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+- **Binary**: Either 0.8 boost or no boost
+- **Reduced impact**: From previous 2.0 to allow more discovery
+
+
+### 4. Trending Score
+```javascript
+trendingScore = TRENDING_COEFFICIENT * getTrendingScore(likes, comments, hoursOld)
+```
+- **Purpose**: Detect and promote viral content (TikTok-style)
+- **Formula**: `(likes + comments * 1.5) / hoursOld / 10`
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/recommendationAlgo/recommendation.md:99]: @sinap-fb Going to take note on this and discuss it during our net one-on-one

File: backend/recommendationAlgo/recommendation.md [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+| `LARGE_CREATOR_FOLLOWER_THRESHOLD` | 500 | Large creators |
+
+
+### Recency Thresholds
+| Threshold | Value | Time Category |
+|-----------|-------|---------------|
+| `VERY_RECENT_HOURS_THRESHOLD` | 1 | Very recent posts |
+| `RECENT_HOURS_THRESHOLD` | 6 | Recent posts |
+| `TODAY_HOURS_THRESHOLD` | 24 | Today's posts |
+| `WEEK_HOURS_THRESHOLD` | 72 | This week's posts |
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/recommendationAlgo/recommendation.md:158]: 3 days creates a tighter recency window that promotes fresher content, creating a natural decay where some contents lose relevance quickly

File: backend/recommendationAlgo/recommendation.js [hunk 7]
```diff
@@ -1,34 +1,113 @@
...
 const topKRecommendations = (videos, k = TOP_K) => {
-    videos.forEach(video => {
-        priorityQueue.push({ ...video, score: videoScore(video)})
-    })
-    const output = []
-    while (!priorityQueue.isEmpty() && output.length < k) {
-        output.push(priorityQueue.pop())
-    }
-    return output
+   const pq = new PriorityQueue((a,b) => a.score > b.score)
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/recommendationAlgo/recommendation.js:100]: maintains sorted order without full array sort

File: backend/routes/postRoutes/postRoutes.js [hunk 8]
```diff
@@ -0,0 +1,465 @@
...
+        })
+    } catch (error) {
+        res.status(500).json({
+            error: "Failed to create post",
+            details: error.message
+        })
+    }
+})
+
+// Like/unlike a post
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/routes/postRoutes/postRoutes.js:159]: For me ,likes are casual, frequent toggles that benefit from a single endpoint, while follows are deliberate relationship changes that warrant separate endpoints for clear intent and different logic

File: backend/routes/postRoutes/postRoutes.js [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+        if (!post) {
+            return res.status(404).json({ error: "Post not found" })
+        }
+
+        if (post.userId !== userId) {
+            return res.status(403).json({ error: "Not authorized to delete this post" })
+        }
+
+        await prisma.post.delete({
+            where: { id: parseInt(postId) }
```
[the remark this answers is not in the document]
Comment (Fnuworsu) [reply backend/routes/postRoutes/postRoutes.js:254]: @sinap-fb It is with `onDelete: "Cascade"` in prisma schema. Thanks for the heads up
